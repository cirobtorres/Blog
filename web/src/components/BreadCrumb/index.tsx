import { Slash } from "lucide-react";
import {
  Breadcrumb as BreadcrumbRoot,
  BreadcrumbItem,
  BreadcrumbLink,
  BreadcrumbList,
  BreadcrumbPage,
  BreadcrumbSeparator,
} from "../Shadcnui/breadcrumb";

const BreadCrumb = ({
  title,
  category,
}: {
  title: string;
  category: Category;
}) => {
  return (
    <BreadcrumbRoot className="mb-4">
      <BreadcrumbList>
        <BreadcrumbItem>
          <BreadcrumbLink href="/">Home</BreadcrumbLink>
        </BreadcrumbItem>
        <BreadcrumbSeparator>
          <Slash />
        </BreadcrumbSeparator>
        <BreadcrumbItem>
          <BreadcrumbLink href="/artigos">Artigos</BreadcrumbLink>
        </BreadcrumbItem>
        <BreadcrumbSeparator>
          <Slash />
        </BreadcrumbSeparator>
        <BreadcrumbPage>
          <BreadcrumbLink href={`/artigos?category=${category.name}`}>
            {category.name}
          </BreadcrumbLink>
        </BreadcrumbPage>
        <BreadcrumbSeparator>
          <Slash />
        </BreadcrumbSeparator>
        <BreadcrumbItem>
          <BreadcrumbPage className="relative text-blog-foreground-highlight after:absolute after:top-[calc(100%_+_4px)] after:left-0 after:w-full after:h-0.5 after:rounded-full after:bg-gradient-to-r after:from-transparent after:to-blog-foreground-highlight">
            {title}
          </BreadcrumbPage>
        </BreadcrumbItem>
      </BreadcrumbList>
    </BreadcrumbRoot>
  );
};

export default BreadCrumb;
