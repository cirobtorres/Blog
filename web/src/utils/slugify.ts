const slugify = (str: string) => {
  return str
    .toLowerCase()
    .normalize("NFD") // Normaliza a string para decompor caracteres acentuados
    .replace(/[\u0300-\u036f]/g, "") // Remove marcas diacríticas (acentos)
    .trim() // Remove espaços extras no início e no final
    .replace(/[^\w\s-]/g, "") // Remove caracteres especiais
    .replace(/[\s_-]+/g, "-") // Substitui espaços e underscores por hifens
    .replace(/^-+|-+$/g, "");
};

export default slugify;
