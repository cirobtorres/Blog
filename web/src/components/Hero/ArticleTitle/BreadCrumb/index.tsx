import { Slash } from "lucide-react";
import {
  Breadcrumb as BreadcrumbRoot,
  BreadcrumbItem,
  BreadcrumbLink,
  BreadcrumbList,
  BreadcrumbPage,
  BreadcrumbSeparator,
} from "../../../Shadcnui/breadcrumb";
import slugify from "@/utils/slugify";

const BreadCrumb = ({ title, topic }: { title: string; topic: Topic }) => {
  return (
    <BreadcrumbRoot id="breadcrumb" data-testid="breadcrumb" className="mb-4">
      <BreadcrumbList>
        <BreadcrumbItem>
          <BreadcrumbLink
            id="breadcrumb-home"
            data-testid="breadcrumb-home"
            href="/"
          >
            Home
          </BreadcrumbLink>
        </BreadcrumbItem>
        <BreadcrumbSeparator>
          <Slash />
        </BreadcrumbSeparator>
        <BreadcrumbItem>
          <BreadcrumbLink
            id="breadcrumb-article"
            data-testid="breadcrumb-article"
            href="/artigos"
          >
            Artigos
          </BreadcrumbLink>
        </BreadcrumbItem>
        <BreadcrumbSeparator>
          <Slash />
        </BreadcrumbSeparator>
        <BreadcrumbPage>
          <BreadcrumbLink
            id="breadcrumb-article-topic"
            data-testid="breadcrumb-article-topic"
            href={`/artigos?topic=${slugify(topic.name)}`}
          >
            {topic.name}
          </BreadcrumbLink>
        </BreadcrumbPage>
        <BreadcrumbSeparator>
          <Slash />
        </BreadcrumbSeparator>
        <BreadcrumbItem>
          <BreadcrumbPage
            id="breadcrumb-article-title"
            data-testid="breadcrumb-article-title"
            className="relative text-blog-foreground-highlight after:absolute after:top-[calc(100%_+_4px)] after:left-0 after:w-full after:h-0.5 after:rounded-full after:bg-gradient-to-r after:from-transparent after:to-blog-foreground-highlight"
          >
            {title}
          </BreadcrumbPage>
        </BreadcrumbItem>
      </BreadcrumbList>
    </BreadcrumbRoot>
  );
};

export default BreadCrumb;
