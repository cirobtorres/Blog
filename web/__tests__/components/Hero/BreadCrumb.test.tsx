import { render, screen } from "@testing-library/react";
import BreadCrumb from "../../../src/components/Hero/ArticleTitle/BreadCrumb";
import slugify from "../../../src/utils/slugify";

describe("BreadCrumb", () => {
  const mockedTitle =
    "Lorem, ipsum dolor sit amet consectetur adipisicing elit.";
  const mockedBreadCrumb = {
    documentId: "11",
    name: "Ciência da Computação",
    slug: "ciencia-da-computacao",
    description: "Descrição da categoria",
  };

  it("renders every link correctly", () => {
    render(<BreadCrumb title={mockedTitle} topic={mockedBreadCrumb} />);
    const linkA = screen.getByTestId("breadcrumb-home");
    const linkB = screen.getByTestId("breadcrumb-article");
    const linkC = screen.getByTestId("breadcrumb-article-topic");

    expect(linkA).toHaveAttribute("href", "/");
    expect(linkB).toHaveAttribute("href", "/artigos");
    expect(linkC).toHaveAttribute(
      "href",
      `/artigos?topic=${slugify(mockedBreadCrumb.name)}`
    );
  });

  it("renders BreadCrumb current page correctly", () => {
    render(<BreadCrumb title={mockedTitle} topic={mockedBreadCrumb} />);
    const currentRoute = screen.getByTestId("breadcrumb-article-title");
    expect(currentRoute).toHaveClass("after:to-blog-foreground-highlight");
    expect(currentRoute).toHaveTextContent(mockedTitle);
  });

  it("matches the snapshot", () => {
    const { asFragment } = render(
      <BreadCrumb title={mockedTitle} topic={mockedBreadCrumb} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
