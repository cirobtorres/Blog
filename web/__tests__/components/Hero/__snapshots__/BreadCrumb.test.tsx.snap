// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BreadCrumb matches the snapshot 1`] = `
<DocumentFragment>
  <nav
    aria-label="breadcrumb"
    class="mb-4"
    data-testid="breadcrumb"
  >
    <ol
      class="flex flex-wrap items-center gap-1.5 break-words text-sm sm:gap-2.5 text-blog-foreground-readable"
    >
      <li
        class="inline-flex items-center gap-1.5"
      >
        <a
          class="transition-colors duration-500 text-blog-foreground-readable hover:text-blog-foreground-readable-hover [&_a]:hover:text-blog-foreground-readable-hover"
          data-testid="breadcrumb-home"
          href="/"
        >
          Home
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:w-3.5 [&>svg]:h-3.5"
        role="presentation"
      >
        <svg
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
      >
        <a
          class="transition-colors duration-500 text-blog-foreground-readable hover:text-blog-foreground-readable-hover [&_a]:hover:text-blog-foreground-readable-hover"
          data-testid="breadcrumb-article"
          href="/artigos"
        >
          Artigos
        </a>
      </li>
      <li
        aria-hidden="true"
        class="[&>svg]:w-3.5 [&>svg]:h-3.5"
        role="presentation"
      >
        <svg
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <span
        aria-current="page"
        aria-disabled="true"
        class="font-normal"
        role="link"
      >
        <a
          class="transition-colors duration-500 text-blog-foreground-readable hover:text-blog-foreground-readable-hover [&_a]:hover:text-blog-foreground-readable-hover"
          data-testid="breadcrumb-article-topic"
          href="/artigos?topic=ciencia-da-computacao"
        >
          Ciência da Computação
        </a>
      </span>
      <li
        aria-hidden="true"
        class="[&>svg]:w-3.5 [&>svg]:h-3.5"
        role="presentation"
      >
        <svg
          class="lucide lucide-slash"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M22 2 2 22"
          />
        </svg>
      </li>
      <li
        class="inline-flex items-center gap-1.5"
      >
        <span
          aria-current="page"
          aria-disabled="true"
          class="font-normal relative text-blog-foreground-highlight after:absolute after:top-[calc(100%_+_4px)] after:left-0 after:w-full after:h-0.5 after:rounded-full after:bg-gradient-to-r after:from-transparent after:to-blog-foreground-highlight"
          data-testid="breadcrumb-article-title"
          role="link"
        >
          Lorem, ipsum dolor sit amet consectetur adipisicing elit.
        </span>
      </li>
    </ol>
  </nav>
</DocumentFragment>
`;
